pkg/rbd/rbd.go

```diff
@@ -27,10 +27,98 @@ import (
 	"k8s.io/utils/exec"
 )
 
-// PluginFolder defines the location of rbdplugin
+/*
+RADOS omaps usage:
+
+This note details how we preserve idempotent nature of create requests and retain the relationship
+between orchestrator (CO) generated Names and plugin generated names for images and snapshots
+
+The implementation uses Ceph RADOS omaps to preserve the relationship between request name and
+generated image (or snapshot) name. There are 4 types of omaps in use,
+- A "csi.volumes.[csi-id]" (or "csi.volumes"+.+CsiInstanceID), we call this the csiVolsDirectory
+  - stores keys named using the CO generated names for volume requests
+  - keys are named "csi.volume."+[CO generated VolName]
+  - Key value contains the RBD image uuid that is created or will be created, for the CO provided
+  name
```

Comment:
Whenever I see applications making raw use of omaps I get nervous. The reason for this is mostly around the de facto limit of ~100,000 k/v pairs per object. Are we confident that a cluster won't have more than 100k volumes?

Reply:
Fair enough :) The initial numbers I discussed with @dillaman were 100K to be exact.

Anyway, as this information is new to me, and we added to omap to support listing (i.e bunch up the csi volumes under an omap to help with listing). My initial solution to the problem (as we can never answer what scale this may reach) would be to remove the `csi.volumes.` and `csi.snaps.` omap, but retain the volname and volid based omaps and keys, as we are no longer intending to support listing RPCs. Thoughts?

(the other knee jerk reaction was, let's create chapters of csi.volumes and use the same, but that depends on if we do not have to maintain counters around number of keys in an omap and can work with errors returned from a omap setomapval instead, that can help understand it is "full")

Reply:
Currently we can scale upto 100K, which is not essentially a hard ceiling. IOW, the updates to the objects may slow down, but not fail at that exact number.


This was discussed, and a possible path for the future is noted in [this](https://github.com/ceph/ceph-csi/pull/312#issuecomment-487949077) comment.

IOW, we decided to continue with the current scale, and in the future if required shard the objects, which is feasible under the current scheme.